test/test_distributions.py

```diff
@@ -1209,23 +1249,99 @@ def test_laplace_shape_tensor_params(self):
+            (Normal(1, 2), Normal(-3, 4)),
+            (Pareto(1, 2), Chi2(3)),
+            (Pareto(1, 2), Exponential(3)),  # This case fails for n = 10000
+            (Pareto(1, 2), Gamma(3, 4)),  # This case fails for n = 10000, 21000
+            (Pareto(1, 3), Normal(-2, 4)),
+            (Uniform(0.25, 0.75), Beta(3, 4)),
+            (Uniform(1, 2), Chi2(3)),
+            (Uniform(1, 2), Exponential(3)),
+            (Uniform(1, 2), Gamma(3, 4)),
+            (Uniform(-1, 2), Gumbel(-3, 4)),
+            (Uniform(-1, 2), Normal(-3, 4)),
+            (Uniform(2, 3), Pareto(1, 4))
+        ]
+
+        self.infinite_examples = [
+            (Beta(1, 2), Uniform(0.25, 1)),
+            (Beta(1, 2), Uniform(0, 0.75)),
+            (Beta(1, 2), Uniform(0.25, 0.75)),
+            (Beta(1, 2), Pareto(1, 2)),
+            (Exponential(1), Beta(2, 3)),
+            (Exponential(1), Pareto(2, 3)),
+            (Exponential(1), Uniform(-2, 3)),
+            (Gamma(1, 2), Beta(3, 4)),
+            (Gamma(1, 2), Pareto(3, 4)),
+            (Gamma(1, 2), Uniform(-3, 4)),
+            (Gumbel(-1, 2), Beta(3, 4)),
+            (Gumbel(-1, 2), Exponential(3)),
+            (Gumbel(-1, 2), Gamma(3, 4)),
+            (Gumbel(-1, 2), Pareto(3, 4)),
+            (Gumbel(-1, 2), Uniform(-3, 4)),
+            (Laplace(-1, 2), Beta(3, 4)),
+            (Laplace(-1, 2), Exponential(3)),
+            (Laplace(-1, 2), Gamma(3, 4)),
+            (Laplace(-1, 2), Pareto(3, 4)),
+            (Laplace(-1, 2), Uniform(-3, 4)),
+            (Normal(-1, 2), Beta(3, 4)),
+            (Normal(-1, 2), Exponential(3)),
+            (Normal(-1, 2), Gamma(3, 4)),
+            (Normal(-1, 2), Pareto(3, 4)),
+            (Normal(-1, 2), Uniform(-3, 4)),
+            (Pareto(2, 1), Exponential(3)),
+            (Pareto(2, 1), Gamma(3, 4)),
+            (Pareto(1, 2), Normal(-3, 4)),
+            (Pareto(1, 2), Pareto(3, 4)),
+            (Uniform(-1, 1), Beta(2, 2)),
+            (Uniform(0, 2), Beta(3, 4)),
+            (Uniform(-1, 2), Beta(3, 4)),
+            (Uniform(-1, 2), Exponential(3)),
+            (Uniform(-1, 2), Gamma(3, 4)),
+            (Uniform(-1, 2), Pareto(3, 4)),
         ]
 
     def test_kl_monte_carlo(self):
         set_rng_seed(0)  # see Note [Randomized statistical tests]
-        for p, q in self.examples:
-            x = p.sample(sample_shape=(10000,))
+        for p, q in self.finite_examples:
+            x = p.sample(sample_shape=(23000,))
             expected = (p.log_prob(x) - q.log_prob(x)).mean(0)
             actual = kl_divergence(p, q)
             message = 'Incorrect KL({}, {}). expected {}, actual {}'.format(
                 type(p).__name__, type(q).__name__, expected, actual)
             self.assertEqual(expected, actual, prec=0.1, message=message)
 
+    def test_kl_infinite(self):
+        for p, q in self.infinite_examples:
+            self.assertTrue((kl_divergence(p, q) == float('inf')), 
```

Comment:
I believe `kl_divergence` is a tensor so you'll need to compare using `.all()`:
```py
(kl_divergence(p, q) == float('inf')).all()
```
Adam will also probably suggest you change this to `self.assertEqual` with `allow_inf=True`, but that seems less transparent to me. I like the version you wrote.

Reply:
In the test cases, the return values are scalars, hence `all` won't work.

Reply:
Hmm, they shouldn't be scalars, `kl_divergence()` should always return `Tensor`s or `Variable`s.